keysync_initiator.go

```diff
@@ -0,0 +1,172 @@
+	theirNonce, err := requestNonce(addr)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Now, create our own nonce.
+	ourNonce, err := newNonce()
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Next, create a key that the remote enclave is going to use to encrypt
+	// its key material.
+	sbKey, err := newSbKey()
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Now create an attestation document containing our nonce, the remote
+	// enclave's nonce, and the key material that they remote enclave is
+	// supposed to use.
+	ourAttDoc, err := attest(theirNonce[:], ourNonce[:], sbKey.Bytes())
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Send our attestation document to the remote enclave, and get theirs in
+	// return.
+	theirAttDoc, err := requestAttDoc(addr, ourAttDoc)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Finally, verify the attestation document and extract the key material.
+	if err := processAttDoc(theirAttDoc, &ourNonce, sbKey, keyMaterial); err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	return nil
+}
+
+// requestNonce requests a nonce from the remote enclave specified by 'addr'.
+func requestNonce(addr string) (nonce, error) {
+	errStr := "failed to fetch nonce from remote enclave"
+
+	endpoint := fmt.Sprintf("%s%s", addr, pathNonce)
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Decode the Base64-encoded nonce.
+	raw, err := base64.StdEncoding.DecodeString(strings.TrimSpace(string(body)))
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	if len(raw) != nonceLen {
+		return nonce{}, errors.New("remote enclave's nonce has incorrect length")
+	}
+	var n nonce
+	copy(n[:], raw)
+	return n, nil
+}
+
+// requestAttDoc takes as input the remote enclave's address (e.g.,
+// <https://example.com>) and our attestation document.  The function then
+// submits our attestation document to the remote enclave, and returns the
+// remote enclave's attestation document.
+func requestAttDoc(addr string, ourAttDoc []byte) ([]byte, error) {
+	errStr := "failed to fetch attestation doc from remote enclave"
+
+	endpoint := fmt.Sprintf("%s%s", addr, pathKeys)
+
+	// Finally, send our attestation document to the remote enclave.  If
+	// everything works out, the remote enclave is going to respond with its
+	// attestation document.
+	b64AttDoc := base64.StdEncoding.EncodeToString(ourAttDoc)
+	resp, err := http.Post(
+		endpoint,
+		"application/octet-stream",
+		bytes.NewBufferString(b64AttDoc),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", errStr, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
```

Comment:
Is there a reasonable length limitation we could enforce here to prevent denial of service?

Reply:
In this case, 5000 bytes should be enough. Fixed in b06d697.